### PR TITLE
Pool objects caught by the catcher

### PR DESCRIPTION
**Pull request by ekrctb**
Addressing the second homework mentioned in my previous PR #11011.

Before this PR, when a hit object is caught by the catcher, a DHO was created and placed on the catcher plate. Caught objects are constantly created and destroyed, and the allocation overhead is significant. Pooling caught objects is desired to reduce stuttering during the gameplay.

However, several DHO logics that are unneeded for a static caught object interfere with the desired pooling logic. For example, we had to use `ApplyCustomState` to workaround an issue of transforms cleared by DHO.

Thus, I decided to create a new class `CaughtObject` for a caught object, not a derived class of DHO.
To reduce code duplication, I extracted common properties of `DrawablePalpableCatchHitObject` and `CaughtObject` to `IHasCatchObjectState` interface.

Visual properties such as `AccentColour` are copied from the DHO to `CaughtObject`, not bound.
A known issue is if the skin is changed then caught objects will keep old skin's combo colors.
 The issue is properties such as `AccentColour` are only present in DHO, not in HOs. I'm currently not sure how to solve this issue. We cannot bind to a DHO because the DHO will be reused for another hit object while the `CaughtObject` is still alive.

I refactored some of the object catching logic of the catcher.
<s>I removed `ScaleContainer` from catch DHO to make the implementation simpler.</s>
Edit: I added it back.

### CODE REVIEW

File: osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+            Alpha = 1;
+            LifetimeStart = double.MinValue;
+            LifetimeEnd = double.MaxValue;
+
+            base.FreeAfterUse();
+        }
+    }
+
+    public class CaughtFruit : CaughtObject, IHasFruitState
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs:65]: Probably should move these to their own files (even for small classes we generally do this as a rule).

File: osu.Game.Rulesets.Catch/UI/Catcher.cs [hunk 2]
```diff
@@ -441,138 +458,123 @@ private void updateState(CatcherAnimationState state)
...
+            return droppedObject;
         }
 
-        private void removeFromPlate(DrawablePalpableCatchHitObject caughtObject, DroppedObjectAnimation animation)
+        private void clearPlate(DroppedObjectAnimation animation)
         {
-            if (!caughtFruitContainer.Remove(caughtObject))
-                throw new InvalidOperationException("Can only drop a caught object on the plate");
+            var caughtObjects = caughtFruitContainer.Children.ToArray();
+            var droppedObjects = caughtObjects.Select(getDroppedObject).ToArray();
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:538]: Not sure about this double copy. Is there a reason to not just immediately `.Select(getDroppedObject)`, therefore allocating just one array?

File: osu.Game.Rulesets.Catch/UI/Catcher.cs [hunk 3]
```diff
@@ -441,138 +458,123 @@ private void updateState(CatcherAnimationState state)
...
+        private void removeFromPlate(CaughtObject caughtObject, DroppedObjectAnimation animation)
         {
-            var originalX = d.X * Scale.X;
-            var startTime = Clock.CurrentTime;
+            var droppedObject = getDroppedObject(caughtObject);
 
-            d.Anchor = Anchor.TopLeft;
-            d.Position = caughtFruitContainer.ToSpaceOfOtherDrawable(d.DrawPosition, droppedObjectTarget);
+            if (!caughtFruitContainer.Remove(caughtObject))
+                throw new InvalidOperationException("Can only drop a caught object on the plate");
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:553]: Not sure if I'm understanding the intent right, would maybe reword to
```suggestion
                throw new InvalidOperationException("Can only drop objects that were previously caught on the plate");
```

File: osu.Game.Rulesets.Catch.Tests/TestSceneCatcherArea.cs [hunk 4]
```diff
@@ -73,7 +73,10 @@ private void createCatcher(float size)
 
             SetContents(() =>
             {
-                var droppedObjectContainer = new Container();
+                var droppedObjectContainer = new Container
+                {
+                    RelativeSizeAxes = Axes.Both
+                };
```
[the remark this answers is not in the document]
Comment (ekrctb) [reply osu.Game.Rulesets.Catch.Tests/TestSceneCatcherArea.cs:79]: I had a bug of not setting `Anchor` of `CaughtObject` when the object is dropped. The bug was not apparent in this visual test when the container is size 0 (not setting `RelativeSizeAxes`) so I added it for the sake of robustness of the visual test.

File: osu.Game.Rulesets.Catch/Objects/Drawables/DrawablePalpableHasCatchHitObject.cs [hunk 5]
```diff
@@ -5,60 +5,49 @@
...
 using osu.Framework.Graphics;
-using osu.Framework.Graphics.Containers;
 using osuTK;
+using osuTK.Graphics;
 
 namespace osu.Game.Rulesets.Catch.Objects.Drawables
 {
-    public abstract class DrawablePalpableCatchHitObject : DrawableCatchHitObject
+    [Cached(typeof(IHasCatchObjectState))]
+    public abstract class DrawablePalpableHasCatchHitObject : DrawableCatchHitObject, IHasCatchObjectState
```
[the remark this answers is not in the document]
Comment (ekrctb) [reply osu.Game.Rulesets.Catch/Objects/Drawables/DrawablePalpableHasCatchHitObject.cs:14]: This is an error from Rider renaming feature. I'll fix it.

File: osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            HyperDash.Value = objectState.HyperDash.Value;
+        }
+
+        protected override void FreeAfterUse()
+        {
+            ClearTransforms();
+
+            Alpha = 1;
+            LifetimeStart = double.MinValue;
+            LifetimeEnd = double.MaxValue;
```
[the remark this answers is not in the document]
Comment (ekrctb) [reply osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs:59]: Lifetime is set by `Expire` when a caught object is dropped. If lifetime is not reset issues will occur.
I removed assignment of `LifetimeStart` because that was not doing anything (I thought the object will be removed when a replay is rewound but it is not, and I added object removing logic to `OnRevertResult`)

File: osu.Game.Rulesets.Catch/Skinning/Default/CatchHitObjectPiece.cs [hunk 7]
```diff
@@ -54,8 +47,8 @@ protected override void LoadComplete()
 
         protected override void Update()
         {
-            if (BorderPiece != null && DrawableHitObject?.HitObject != null)
-                BorderPiece.Alpha = (float)Math.Clamp((DrawableHitObject.HitObject.StartTime - Time.Current) / 500, 0, 1);
+            if (BorderPiece != null && ObjectState?.HitObject != null)
```
[the remark this answers is not in the document]
Comment (ekrctb) [reply osu.Game.Rulesets.Catch/Skinning/Default/CatchHitObjectPiece.cs:50]: For a DHO, `HitObject` can be null here for some reason.

File: osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ekrctb) [reply osu.Game.Rulesets.Catch/Objects/Drawables/CaughtObject.cs:59]: I somehow thought lifetime had to reset but apparently it is reset by `DrawablePool` and the lines above is not necessary. I think I misunderstood a different bug (that I solved).

File: osu.Game.Rulesets.Catch/UI/Catcher.cs (hunk 3 above)
Comment (ekrctb) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:553]: Actually this check is not needed anymore because this method was public but it is now private and the exception shouldn't occur.

**Review by bdach — DISMISSED** [review]
Looks good to go as far as I can tell :+1:

**Comment by bdach** [comment]
I think the accent colour issue can be resolved in a separate pull. An option would be to have `AccentColour` in the hitobjects themselves and associate `CaughtObject`s with them, but I imagine managing that state could turn ugly...

File: osu.Game.Rulesets.Catch/Objects/Drawables/DrawableBanana.cs [hunk 8]
```diff
@@ -44,12 +44,12 @@ protected override void UpdateInitialTransforms()
             const float end_scale = 0.6f;
             const float random_scale_range = 1.6f;
 
-            ScaleContainer.ScaleTo(HitObject.Scale * (end_scale + random_scale_range * RandomSingle(3)))
-                          .Then().ScaleTo(HitObject.Scale * end_scale, HitObject.TimePreempt);
+            this.ScaleTo(HitObject.Scale * (end_scale + random_scale_range * RandomSingle(3)))
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/Objects/Drawables/DrawableBanana.cs:47]: This is probably not what we want to be doing. The rationale behind not adjusting one's own scale is that mods may want to do so as well, causing potentially loss (or otherwise incorrect) transforms.

Can you add back the local scale container and instead forward the interface implementation to the `HitObject` model's scale (can be done implicitly to avoid conflict with the underlying `Drawable.Scale`)?

File: osu.Game.Rulesets.Catch/Objects/Drawables/DrawableDroplet.cs [hunk 9]
```diff
@@ -26,9 +24,9 @@ public DrawableDroplet([CanBeNull] CatchHitObject h)
         [BackgroundDependencyLoader]
         private void load()
         {
-            ScaleContainer.Child = new SkinnableDrawable(
+            AddInternal(new SkinnableDrawable(
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/Objects/Drawables/DrawableDroplet.cs:27]: As above.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by ekrctb** [comment]
My commit above changed the radius of banana considered for the caught object stacking. It is now more correct and matches visual (previously scaling transform was ignored for the radius).

File: osu.Game.Rulesets.Catch/UI/Catcher.cs [hunk 10]
```diff
@@ -124,7 +134,11 @@ public Catcher([NotNull] Container trailsTarget, [NotNull] Container droppedObje
             InternalChildren = new Drawable[]
             {
                 hitExplosionPool = new DrawablePool<HitExplosion>(10),
-                caughtFruitContainer = new Container<DrawablePalpableCatchHitObject>
+                caughtFruitPool = new DrawablePool<CaughtFruit>(50),
```
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:137]: As a safety to avoid asynchronous loading, we'd usually put these inside the `[BackgroundDependencyLoader]` method, but I've checked this case is fine here (async happens from `CreateDrawableRuleset` and includes all `ctor`s).

**Review by peppy — APPROVED** [review]
Looks good here!
